crates/integrations/datafusion/src/physical_plan/repartition.rs

```diff
@@ -0,0 +1,966 @@
+///
+/// ## Fallback
+///
+/// If no suitable hash columns are found, falls back to round-robin batch partitioning
+/// to ensure even load distribution across partitions.
+fn determine_partitioning_strategy(
+    input: &Arc<dyn ExecutionPlan>,
+    table_metadata: &TableMetadata,
+    target_partitions: NonZeroUsize,
+) -> DFResult<Partitioning> {
+    let partition_spec = table_metadata.default_partition_spec();
+    let table_schema = table_metadata.current_schema();
+    let input_schema = input.schema();
+    let target_partition_count = target_partitions.get();
+
+    // Check if partition spec has transforms suitable for hash partitioning
+    let has_hash_friendly_transforms = partition_spec
+        .fields()
+        .iter()
+        .any(|pf| matches!(pf.transform, Transform::Identity | Transform::Bucket(_)));
+
+    let partition_col_result = input_schema.index_of(PARTITION_VALUES_COLUMN);
+    let is_partitioned_table = !partition_spec.is_unpartitioned();
+
+    match (is_partitioned_table, partition_col_result) {
+        // Case 1: Partitioned table with _partition column present
+        (true, Ok(partition_col_idx)) => {
+            let partition_field = input_schema.field(partition_col_idx);
+            if partition_field.name() != PARTITION_VALUES_COLUMN {
+                return Err(DataFusionError::Plan(format!(
+                    "Expected {} column at index {}, but found '{}'",
+                    PARTITION_VALUES_COLUMN,
+                    partition_col_idx,
+                    partition_field.name()
+                )));
+            }
+
+            let partition_expr = Arc::new(Column::new(PARTITION_VALUES_COLUMN, partition_col_idx))
+                as Arc<dyn PhysicalExpr>;
+
+            return if has_hash_friendly_transforms {
+                Ok(Partitioning::Hash(
+                    vec![partition_expr],
+                    target_partition_count,
+                ))
+            } else {
+                Ok(Partitioning::RoundRobinBatch(target_partition_count))
+            };
+        }
+
+        // Case 2: Partitioned table missing _partition column (warning)
+        (true, Err(_)) => {
+            tracing::warn!(
+                "Partitioned table input missing {} column. \
+                 Consider adding partition projection before repartitioning.",
+                PARTITION_VALUES_COLUMN
+            );
+        }
+
+        // Case 3: Unpartitioned table with _partition column
+        (false, Ok(_)) => {
+            tracing::warn!(
+                "Input contains {} column but table is unpartitioned. \
+                 This may indicate unnecessary projection.",
+                PARTITION_VALUES_COLUMN
+            );
+        }
+
+        // Case 4: Unpartitioned table without _partition column
+        (false, Err(_)) => {
+            // Nothing to do - fall through to source column analysis
+        }
+    }
+
+    let hash_column_names: Vec<&str> = partition_spec
```

Comment:
Why we need this part? The physical plan is generated by `IcebergTableProvider`, so we could ensure that the `_partition` column will be inserted.